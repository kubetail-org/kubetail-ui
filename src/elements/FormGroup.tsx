// Copyright 2024 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createContext, forwardRef, useMemo } from 'react';

import { cn } from '@/lib/utils';

type ContextState = {
  controlId: string | undefined,
};

const Context = createContext({} as ContextState); // for passing `controlId`
const baseCls = 'space-y-1';

interface Props extends React.ComponentPropsWithoutRef<'div'> {
  controlId?: string;
}

const FormGroup = forwardRef((
  {
    className,
    controlId,
    ...props
  }: Props,
  ref: React.ForwardedRef<HTMLDivElement>,
) => {
  const context = useMemo(() => ({ controlId }), [controlId]);

  return (
    <Context.Provider value={context}>
      <div
        {...props}
        ref={ref}
        className={cn(baseCls, className)}
      />
    </Context.Provider>
  );
});

FormGroup.displayName = 'FormGroup';

export { Context };
export default FormGroup;
