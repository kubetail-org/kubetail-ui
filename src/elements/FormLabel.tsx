// Copyright 2024 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { forwardRef, useContext } from 'react';

import { cn } from '@/lib/utils';

import { Context as FormGroupContext } from './FormGroup';

const baseCls = 'text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70';

interface Props extends React.ComponentPropsWithoutRef<'label'> {
  as?: React.ElementType;
  htmlFor?: string;
}

const FormLabel = forwardRef((
  {
    as = 'label',
    className,
    htmlFor,
    ...props
  }: Props,
  ref: React.ForwardedRef<HTMLLabelElement>,
) => {
  const Tag = as;
  const { controlId } = useContext(FormGroupContext);

  return (
    <Tag
      {...props}
      ref={ref}
      htmlFor={htmlFor || controlId}
      className={cn(baseCls, className)}
    />
  );
});

FormLabel.displayName = 'FormLabel';

export default FormLabel;
