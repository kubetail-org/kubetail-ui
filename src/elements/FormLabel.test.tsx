// Copyright 2024 Andres Morey
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { render } from '@testing-library/react';

import FormGroup from './FormGroup';
import FormLabel from './FormLabel';

describe('FormLabel', () => {
  it('renders contents properly', () => {
    const { asFragment } = render(<FormLabel>contents</FormLabel>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('appends custom className', () => {
    const { asFragment } = render(<FormLabel className="extra-class" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('inherits id from FormGroup.controlId', () => {
    const { asFragment } = render(
      <FormGroup controlId="input-id">
        <FormLabel />
      </FormGroup>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
