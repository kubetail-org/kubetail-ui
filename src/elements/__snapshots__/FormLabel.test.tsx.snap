// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormLabel > appends custom className 1`] = `
<DocumentFragment>
  <label
    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 extra-class"
  />
</DocumentFragment>
`;

exports[`FormLabel > inherits id from FormGroup.controlId 1`] = `
<DocumentFragment>
  <div
    class="space-y-1"
  >
    <label
      class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
      for="input-id"
    />
  </div>
</DocumentFragment>
`;

exports[`FormLabel > renders contents properly 1`] = `
<DocumentFragment>
  <label
    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
  >
    contents
  </label>
</DocumentFragment>
`;
